### PR TITLE
Missing runtime store

### PR DESCRIPTION
**Pull request by ramarag**
This 	annotates filtered packages in deps.json, to give a better message if the runtimeStore is missing

@livarcocc @eerhardt @gkhanna79 PTAL

### CODE REVIEW

**Comment by ramarag** [comment]
@dsplaisted @nguerrera @livarcocc any idea why only ubuntu legs are only failing ?

The error in the CI is
```
/mnt/j/workspace/dotnet_sdk/master/debug_ubuntu16.04_prtest/bin/Debug/Tests/RunFromOutputFolderConflicts/RunFromOutputFolderConflicts/bin/Debug/netcoreapp2.0/RunFromOutputFolderConflicts.dll
15:28:53 /mnt/j/workspace/dotnet_sdk/master/debug_ubuntu16.04_prtest/build/build.proj(169,5): error : Error: assembly specified in the dependencies manifest was not found -- package: 'System.Runtime.InteropServices.RuntimeInformation', version: '4.3.0', path: 'runtimes/unix/lib/netstandard1.1/System.Runtime.InteropServices.RuntimeInformation.dll'
```

**Comment by ramarag** [comment]
fixes https://github.com/dotnet/sdk/issues/1042

**Comment by livarcocc** [comment]
cc @dsplaisted who is investigating this class of errors right now.

File: src/Tasks/Microsoft.NET.Build.Tasks.UnitTests/all.asset.types.osx.deps.json [hunk 1]
```diff
@@ -18,15 +18,15 @@
           "all.asset.types.dll": {}
         }
       },
-      "libuv/1.9.0": {
+      "Libuv/1.9.0": {
```
Comment (gkhanna79) [inline src/Tasks/Microsoft.NET.Build.Tasks.UnitTests/all.asset.types.osx.deps.json]: Why was this change required?

File: src/Tasks/Microsoft.NET.Build.Tasks.UnitTests/all.asset.types.osx.deps.json [hunk 2]
```diff
@@ -18,15 +18,15 @@
...
         "native": {
           "runtimes/osx/native/libuv.dylib": {}
         }
       },
-      "microsoft.netcore.app/1.0.0": {},
-      "sharedcontenta/1.0.0": {},
-      "system.runtime.analyzers/1.1.0": {},
-      "system.spatial/5.7.0": {
+      "Microsoft.NETCore.App/1.0.0": {},
```
Comment (gkhanna79) [inline src/Tasks/Microsoft.NET.Build.Tasks.UnitTests/all.asset.types.osx.deps.json]: Likewise, why do we need these casing changes?

File: src/Tasks/Microsoft.NET.Build.Tasks/DependencyContextBuilder.cs [hunk 3]
```diff
@@ -25,6 +26,7 @@ internal class DependencyContextBuilder
         private IEnumerable<string> _privateAssetPackageIds;
         private CompilationOptions _compilationOptions;
         private string _referenceAssembliesPath;
+        private Dictionary<PackageIdentity, StringBuilder> _packagesThatWhereFiltered;
```
Comment (gkhanna79) [inline src/Tasks/Microsoft.NET.Build.Tasks/DependencyContextBuilder.cs]: _filteredPackages?

**Review by gkhanna79 — COMMENTED** [review]
I will let @eerhardt review the overall change.

My other feedback is to please add comments to the code.

File: src/Tasks/Microsoft.NET.Build.Tasks.UnitTests/all.asset.types.osx.deps.json (hunk 1 above)
Comment (eerhardt) [inline src/Tasks/Microsoft.NET.Build.Tasks.UnitTests/all.asset.types.osx.deps.json]: These casing changes are caused from using the new DependencyModel version that doesn't have all the "ToLower" hacks in it to work around the NuGet breaking change for netcoreapp1.0.

---
In reply to: [111397750](https://github.com/dotnet/sdk/pull/1119#discussion_r111397750) [](ancestors = 111397750)

File: test/Microsoft.NET.Publish.Tests/GivenThatWeWantToPublishAProjectWithDependencies.cs [hunk 4]
```diff
@@ -147,9 +149,11 @@ public void It_publishes_projects_with_simple_dependencies_with_filter_profile()
...
 
            var runtimeConfig = ReadJson(System.IO.Path.Combine(publishDirectory.ToString(), $"{project}.runtimeconfig.json"));
-
            runtimeConfig["runtimeOptions"]["tfm"].ToString().Should().Be("netcoreapp1.1");
-            
+
+            var depsJson = ReadJson(System.IO.Path.Combine(publishDirectory.ToString(), $"{project}.deps.json"));
+            depsJson["libraries"]["Newtonsoft.Json/9.0.1"]["runtimeStoreManifestName"].ToString().Should().Be($"{manifestFileName1} {manifestFileName2}");
+
 //TODO: Enable testing the run once dotnet host has the notion of looking up shared packages
```
Comment (eerhardt) [inline test/Microsoft.NET.Publish.Tests/GivenThatWeWantToPublishAProjectWithDependencies.cs]: This should be true now right?  Can we get an issue logged for this TODO so we do this testing?

File: test/Microsoft.NET.Publish.Tests/GivenThatWeWantToPublishAProjectWithDependencies.cs [hunk 5]
```diff
@@ -147,9 +149,11 @@ public void It_publishes_projects_with_simple_dependencies_with_filter_profile()
             });
 
            var runtimeConfig = ReadJson(System.IO.Path.Combine(publishDirectory.ToString(), $"{project}.runtimeconfig.json"));
-
            runtimeConfig["runtimeOptions"]["tfm"].ToString().Should().Be("netcoreapp1.1");
-            
+
+            var depsJson = ReadJson(System.IO.Path.Combine(publishDirectory.ToString(), $"{project}.deps.json"));
+            depsJson["libraries"]["Newtonsoft.Json/9.0.1"]["runtimeStoreManifestName"].ToString().Should().Be($"{manifestFileName1} {manifestFileName2}");
```
Comment (eerhardt) [inline test/Microsoft.NET.Publish.Tests/GivenThatWeWantToPublishAProjectWithDependencies.cs]: Should have an end-to-end test that ensures when this package can't be found (because the runtime store is unavailable), that a good exception is displayed.

File: src/Tasks/Microsoft.NET.Build.Tasks/GenerateDepsFile.cs [hunk 6]
```diff
@@ -104,13 +109,43 @@ protected override void ExecuteCore()
                 PlatformLibraryName,
                 IsSelfContained);
 
+            Dictionary<PackageIdentity, StringBuilder> packagesThatWhereFiltered = null;
```
Comment (eerhardt) [inline src/Tasks/Microsoft.NET.Build.Tasks/GenerateDepsFile.cs]: This chunk of logic should be encapsulated in a method to enable easier reading.

File: src/Tasks/Microsoft.NET.Build.Tasks/DependencyContextBuilder.cs [hunk 7]
```diff
@@ -185,6 +193,40 @@ private static string GenerateRuntimeSignature(IEnumerable<LockFileTargetLibrary
...
+            IEnumerable<Dependency> dependencies,
+            bool serviceable,
+            string path = null,
+            string hashPath = null
+            )
+        {
+            string runtimeStoreManifestName = null;
+            var pkg = new PackageIdentity(name, NuGetVersion.Parse(version));
+            StringBuilder listofManifests;
+            if (_packagesThatWhereFiltered != null && _packagesThatWhereFiltered.TryGetValue(pkg, out listofManifests))
```
Comment (eerhardt) [inline src/Tasks/Microsoft.NET.Build.Tasks/DependencyContextBuilder.cs]: Could use:

`_packagesThatWhereFiltered?.TryGetValue(pkg, out listofManifests) == true` instead.

File: src/Tasks/Microsoft.NET.Build.Tasks/GenerateDepsFile.cs [hunk 8]
```diff
@@ -104,13 +109,43 @@ protected override void ExecuteCore()
...
+                    var packagesSpecified = StoreArtifactParser.Parse(targetManifestFile);
+                    var targetManifestFileName = Path.GetFileName(targetManifestFile);
+
+                    foreach (var pkg in packagesSpecified)
+                    {
+                        Log.LogMessage(MessageImportance.Low, string.Format(CultureInfo.CurrentCulture, Strings.PackageInfoLog, pkg.Id, pkg.Version));
+                        StringBuilder fileList;
+                        if (packagesThatWhereFiltered.TryGetValue(pkg, out fileList))
+                        {
+                            fileList.Append(" ");
```
Comment (eerhardt) [inline src/Tasks/Microsoft.NET.Build.Tasks/GenerateDepsFile.cs]: We should use a `;` separator here, so we can tell the difference between a file name with space in it, and two separate files.

File: src/Tasks/Microsoft.NET.Build.Tasks/build/Microsoft.NET.Publish.targets [hunk 9]
```diff
@@ -36,6 +36,10 @@ Copyright (c) .NET Foundation. All rights reserved.
     </ResolvedFileToPublish>
   </ItemDefinitionGroup>
 
+  <ItemGroup>
```
Comment (eerhardt) [inline src/Tasks/Microsoft.NET.Build.Tasks/build/Microsoft.NET.Publish.targets]: I'm not crazy about putting this in between the "Publish" comment above and the "Publish" target below.  I see there is already "stuff" in between them, but adding to this makes it messy, IMO.

File: test/Microsoft.NET.Publish.Tests/GivenThatWeWantToPublishAProjectWithDependencies.cs (hunk 5 above)
Comment (ramarag) [inline test/Microsoft.NET.Publish.Tests/GivenThatWeWantToPublishAProjectWithDependencies.cs]: that is currently being done in cli. Has SDK enabled first class test experience for .Netcore 2.0 ?

Comment (eerhardt) [inline test/Microsoft.NET.Publish.Tests/GivenThatWeWantToPublishAProjectWithDependencies.cs]: @livarcocc @dsplaisted - ?

Comment (eerhardt) [inline test/Microsoft.NET.Publish.Tests/GivenThatWeWantToPublishAProjectWithDependencies.cs]: Looks like you are talking about https://github.com/dotnet/cli/blob/master/test/dotnet-store.Tests/GivenDotnetStoresAndPublishesProjects.cs#L101-L104.

As long as we ensure this test fails when these changes are merged, then I'm fine with continuing to test the end-to-end in the cli.

File: src/Tasks/Microsoft.NET.Build.Tasks/DependencyContextBuilder.cs [hunk 10]
```diff
@@ -185,6 +193,40 @@ private static string GenerateRuntimeSignature(IEnumerable<LockFileTargetLibrary
             return dependencies;
         }
 
+        private RuntimeLibrary CreateRuntimeLibrary(string type,
```
Comment (livarcocc) [inline src/Tasks/Microsoft.NET.Build.Tasks/DependencyContextBuilder.cs]: The formatting style is varying here. Please, when you put an argument on a newline, put them all in a new line. So, string type should be on the line below.

Also, the closing parenthesis should be in the same line as string hashPath = null

File: src/Tasks/Microsoft.NET.Build.Tasks/DependencyContextBuilder.cs [hunk 11]
```diff
@@ -185,6 +193,40 @@ private static string GenerateRuntimeSignature(IEnumerable<LockFileTargetLibrary
...
+        {
+            string runtimeStoreManifestName = null;
+            var pkg = new PackageIdentity(name, NuGetVersion.Parse(version));
+            StringBuilder listofManifests;
+            if (_packagesThatWhereFiltered != null && _packagesThatWhereFiltered.TryGetValue(pkg, out listofManifests))
+            {
+                runtimeStoreManifestName = listofManifests.ToString();
+            }
+
+            return new RuntimeLibrary(type,
```
Comment (livarcocc) [inline src/Tasks/Microsoft.NET.Build.Tasks/DependencyContextBuilder.cs]: Same, type should be in its own line.

File: src/Tasks/Microsoft.NET.Build.Tasks/GenerateDepsFile.cs (hunk 8 above)
Comment (livarcocc) [inline src/Tasks/Microsoft.NET.Build.Tasks/GenerateDepsFile.cs]: Also, how about fileList.Append($";{targetManifestFileName}")

File: src/Tasks/Microsoft.NET.Build.Tasks/build/Microsoft.NET.Publish.targets (hunk 9 above)
Comment (ramarag) [inline src/Tasks/Microsoft.NET.Build.Tasks/build/Microsoft.NET.Publish.targets]: moved the comment below

File: test/Microsoft.NET.Publish.Tests/GivenThatWeWantToPublishAProjectWithDependencies.cs (hunk 4 above)
Comment (ramarag) [inline test/Microsoft.NET.Publish.Tests/GivenThatWeWantToPublishAProjectWithDependencies.cs]: will track in https://github.com/dotnet/sdk/issues/1036#issuecomment-294038564

File: src/Tasks/Microsoft.NET.Build.Tasks/GenerateDepsFile.cs (hunk 6 above)
Comment (ramarag) [inline src/Tasks/Microsoft.NET.Build.Tasks/GenerateDepsFile.cs]: done

**Comment by ramarag** [comment]
@eerhardt @livarcocc any other concerns ?

**Review by eerhardt — APPROVED** [review]
